things/policies/service.go

```diff
@@ -143,6 +156,31 @@ func (svc service) AddPolicy(ctx context.Context, token string, p Policy) (Polic
 	return Policy{}, errors.ErrAuthorization
 }
 
+// checkSubject is used to check if the subject is valid.
+//
+// 1. If the subject is a thing (external is false), check the following:
+//   - The user is the owner of the thing
+//   - The user has `c_share` action on the thing
+//
+// 2. If the subject is a user (external is true), check the following:
+//   - The user is the owner of the user
+//   - The user has `c_share` action on the user
+func (svc service) checkSubject(ctx context.Context, userID string, external bool, p Policy) error {
+	switch external {
+	case false:
+		ar := AccessRequest{Subject: userID, Object: p.Subject, Action: sharePolicyAction}
+		if _, err := svc.policies.EvaluateThingAccess(ctx, ar); err != nil {
+			return err
+		}
+	case true:
+		if err := svc.usersAuthorize(ctx, userID, p.Subject, sharePolicyAction, ClientEntityType); err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
```suggestion
switch {
case external:
	ar := AccessRequest{Subject: userID, Object: p.Subject, Action: sharePolicyAction}
	if _, err := svc.policies.EvaluateThingAccess(ctx, ar); err != nil {
		return err
	}
default:
	if err := svc.usersAuthorize(ctx, userID, p.Subject, sharePolicyAction, ClientEntityType); err != nil {
		return err
	}
}

return nil
```
This looks a little bit more elegant. Or use simple `if` instead.